### PR TITLE
[AIRFLOW-6072] aws_hook: Outbound http proxy setting and other enhancements

### PR DESCRIPTION
**Pull request by baolsen**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

AIRFLOW-6072 and AIRFLOW-6075

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Additional capabilities on aws_hook with documentation to match. These capabilities allow for more complex scenarios such as:

1. Specifying boto3.client config, used eg when connecting behind http_proxy
1. Additional logging to help understand how credentials are found and used
1. kwargs for assume_role provide more flexibility
1. Allow to specify boto3.session parameters such as 'profile' which thereby supports ~/.aws/config files

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

I manually tested the new capabilities against my AWS account
### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

Modernised the aws hook documentation and linked it up to the Class.

### CODE REVIEW

File: airflow/contrib/hooks/aws_hook.py [hunk 1]
```diff
@@ -18,79 +18,38 @@
...
 import logging
 
 import boto3
+from botocore.config import Config
 
 from airflow.exceptions import AirflowException
 from airflow.hooks.base_hook import BaseHook
 
 
-def _parse_s3_config(config_file_name, config_format='boto', profile=None):
```
Comment (baolsen) [inline airflow/contrib/hooks/aws_hook.py:33]: Moved this function after the Class, so the documentation for actual AWS Hook is at the top of the page in Sphinx

**Comment by baolsen** [comment]
This build is failing due to some GCP error but none of the files I changes are GCP related.

Main Airflow build is currently failing, I probably rebased from there so my build is failing. Will wait until the main Airflow build passes and then rebase this PR, which should fix this build.

**Comment by potiuk** [comment]
It looked like it was a flaky test. I restarted just this job that failed. Stabilising such flaky tests is sometimes tricky.

However. we are just about to merge a change to use pytests rather than nosetests as test execution engine (hopefully today @nuclearpinguin 🙏 ). So this will happen less often we hope.

This will enable us to mark some tests as flaky and have pytest retry such flaky tests in case they fail (and we will continue fixing  such flaky tests in parallel).

**Comment by baolsen** [comment]
Thanks @potiuk that makes sense. My build passed now, so likely a flaky test as you said :)

**Review by potiuk — APPROVED** [review]
NICE! Finally AWS gets some love in Airflow :).

**Pull request merged**